File: presence.go
package main

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"time"
)

var (
	name     = "anon"
	peers    = map[string]bool{}
	presence = map[string][2]any{} // name -> [t, addr]
	secret   []byte                // HMAC key, generated once at startup
)

const ipUDPHeader = 28

// tokenFor derives a per-address token via HMAC — no per-peer state to store.
// See README: "You could use a hash of their address and a secret."
func tokenFor(addr string) string {
	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(addr))
	return hex.EncodeToString(mac.Sum(nil))[:32]
}

func splitAddr(s string) (string, int) {
	i := strings.LastIndex(s, ":")
	p, _ := strconv.Atoi(s[i+1:])
	return s[:i], p
}

func main() {
	if len(os.Args) > 1 {
		name = os.Args[1]
	} else if v := os.Getenv("NAME"); v != "" {
		name = v
	}

	secret = make([]byte, 32)
	rand.Read(secret)

	conn, _ := net.ListenPacket("udp4", ":24254")
	defer conn.Close()

	sendTo := func(host string, port int, msgs []any) {
		addr, _ := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:%d", host, port))
		b, _ := json.Marshal(msgs)
		conn.WriteTo(b, addr)
	}
	sendAddr := func(addr net.Addr, msgs []any) {
		b, _ := json.Marshal(msgs)
		conn.WriteTo(b, addr)
	}

	for _, b := range [][2]string{{"148.71.89.128", "24254"}, {"159.69.54.127", "24254"}} {
		port, _ := strconv.Atoi(b[1])
		key := b[0] + ":" + b[1]
		peers[key] = true
		sendTo(b[0], port, []any{
			map[string]any{"PleaseSendPeers": map[string]any{}},
			map[string]any{"PleaseAlwaysReturnThisMessage": map[string]any{"cookie": tokenFor(key)}},
		})
	}

	go func() {
		for range time.Tick(5 * time.Second) {
			t := time.Now().Unix()
			for p := range peers {
				h, port := splitAddr(p)
				sendTo(h, port, []any{
					map[string]any{"IAmHere": map[string]any{"name": name, "t": t}},
					map[string]any{"PleaseAlwaysReturnThisMessage": map[string]any{"cookie": tokenFor(p)}},
				})
			}
		}
	}()

	go func() {
		for range time.Tick(10 * time.Second) {
			now := time.Now().Unix()
			fmt.Printf("\n--- Who is here (%s) ---\n", name)
			for n, info := range presence {
				t, _ := info[0].(int64)
				addr, _ := info[1].(string)
				fmt.Printf("  %s  %s  %ds ago\n", n, addr, now-t)
			}
		}
	}()

	buf := make([]byte, 65535)
	for {
		n, addr, err := conn.ReadFrom(buf)
		if err != nil {
			continue
		}
		data := make([]byte, n)
		copy(data, buf[:n])
		var msgs []map[string]any
		if json.Unmarshal(data, &msgs) != nil {
			continue
		}
		src := addr.String()
		peers[src] = true
		reqBytes := float64(len(data) + ipUDPHeader)
		now := time.Now().Unix()

		// Verified if this packet echoes the HMAC token we would compute for this address.
		isVerified := false
		for _, m := range msgs {
			if ar, ok := m["AlwaysReturned"].(map[string]any); ok {
				if tok, ok := ar["cookie"].(string); ok && tok == tokenFor(src) {
					isVerified = true
					break
				}
			}
		}

		var out []any
		var theirToken any

		for _, m := range msgs {
			if p, ok := m["Peers"].(map[string]any); ok {
				if ps, ok := p["peers"].([]any); ok {
					for _, peer := range ps {
						if s, ok := peer.(string); ok {
							peers[s] = true
						}
					}
				}
			}
			if iah, ok := m["IAmHere"].(map[string]any); ok {
				nm, _ := iah["name"].(string)
				t, _ := iah["t"].(float64)
				presence[nm] = [2]any{int64(t), src}
			}
			if hiw, ok := m["HereIsWho"].(map[string]any); ok {
				if ns, ok := hiw["nodes"].([]any); ok {
					for _, node := range ns {
						if nm, ok := node.(map[string]any); ok {
							n, _ := nm["name"].(string)
							t, _ := nm["t"].(float64)
							a, _ := nm["addr"].(string)
							if a == "" {
								a = src
							}
							presence[n] = [2]any{int64(t), a}
						}
					}
				}
			}
			if _, ok := m["WhoIsHere"]; ok {
				cutoff := now - 60
				var nodes []any
				for n, info := range presence {
					t, _ := info[0].(int64)
					a, _ := info[1].(string)
					if t >= cutoff {
						nodes = append(nodes, map[string]any{"name": n, "t": t, "addr": a})
					}
				}
				out = append(out, map[string]any{"HereIsWho": map[string]any{"nodes": nodes}})
			}
			if _, ok := m["PleaseSendPeers"]; ok {
				var pl []string
				for p := range peers {
					pl = append(pl, p)
					if len(pl) >= 20 {
						break
					}
				}
				out = append(out, map[string]any{"Peers": map[string]any{"peers": pl}})
			}
			if patm, ok := m["PleaseAlwaysReturnThisMessage"].(map[string]any); ok {
				theirToken = patm["cookie"]
			}
		}

		if len(out) > 0 {
			if theirToken != nil {
				out = append(out, map[string]any{"AlwaysReturned": map[string]any{"cookie": theirToken}})
			}
			// Include our token in every reply so an unverified peer can echo it back
			// and receive full responses starting from the very next exchange.
			out = append(out, map[string]any{"PleaseAlwaysReturnThisMessage": map[string]any{"cookie": tokenFor(src)}})

			if !isVerified {
				payload, _ := json.Marshal(out)
				if float64(len(payload)) > 2.5*reqBytes {
					// Minimum: our token (lets them bootstrap verification) + at most 1 peer.
					trimmed := []any{map[string]any{"PleaseAlwaysReturnThisMessage": map[string]any{"cookie": tokenFor(src)}}}
					for _, item := range out {
						if pm, ok := item.(map[string]any); ok {
							if p, ok := pm["Peers"].(map[string]any); ok {
								if pl, ok := p["peers"].([]string); ok && len(pl) > 0 {
									trimmed = append(trimmed, map[string]any{"Peers": map[string]any{"peers": pl[:1]}})
								}
								break
							}
						}
					}
					out = trimmed
				}
			}
			sendAddr(addr, out)
		}
	}
}

File: presence.py
import socket, json, time, sys, os, hmac as _hmac, hashlib, secrets

NAME = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("NAME", "anon")
BOOTSTRAP = [("148.71.89.128", 24254), ("159.69.54.127", 24254)]
IP_UDP_HEADER = 28

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(("0.0.0.0", 24254))
sock.settimeout(1.0)

peers = set()
presence = {}  # name -> {"t": unix_seconds, "addr": "ip:port"}

# HMAC-derived token: no per-peer state to store.
# See README: "You could use a hash of their address and a secret."
secret = secrets.token_bytes(32)

def token_for(addr_str):
    return _hmac.new(secret, addr_str.encode(), hashlib.sha256).hexdigest()[:32]

def send(addr, msgs):
    sock.sendto(json.dumps(msgs).encode(), addr)

def split(s):
    i = s.rfind(":")
    return s[:i], int(s[i+1:])

for b in BOOTSTRAP:
    key = f"{b[0]}:{b[1]}"
    peers.add(key)
    send(b, [{"PleaseSendPeers": {}}, {"PleaseAlwaysReturnThisMessage": {"cookie": token_for(key)}}])

last_ping = last_print = 0

while True:
    now = time.time()

    if now - last_ping > 5:
        for p in list(peers):
            send(split(p), [{"IAmHere": {"name": NAME, "t": int(now)}},
                            {"PleaseAlwaysReturnThisMessage": {"cookie": token_for(p)}}])
        last_ping = now

    if now - last_print > 10:
        print(f"\n--- Who is here ({NAME}) ---")
        for n, info in sorted(presence.items()):
            print(f"  {n}  {info['addr']}  {int(now - info['t'])}s ago")
        last_print = now

    try:
        data, addr = sock.recvfrom(65535)
    except socket.timeout:
        continue

    try:
        msgs = json.loads(data.decode())
    except Exception:
        continue

    src = f"{addr[0]}:{addr[1]}"
    peers.add(src)
    req_bytes = len(data) + IP_UDP_HEADER

    # Verified if this packet echoes the HMAC token we would compute for this address.
    is_verified = any(
        isinstance(m.get("AlwaysReturned"), dict) and
        m["AlwaysReturned"].get("cookie") == token_for(src)
        for m in msgs if "AlwaysReturned" in m
    )

    out = []
    their_token = None

    for m in msgs:
        if "Peers" in m:
            for p in m["Peers"].get("peers", []):
                peers.add(p)
        if "IAmHere" in m:
            presence[m["IAmHere"]["name"]] = {"t": m["IAmHere"]["t"], "addr": src}
        if "HereIsWho" in m:
            for node in m["HereIsWho"].get("nodes", []):
                presence[node["name"]] = {"t": node["t"], "addr": node.get("addr", src)}
        if "WhoIsHere" in m:
            cutoff = now - 60
            out.append({"HereIsWho": {"nodes": [
                {"name": n, "t": info["t"], "addr": info["addr"]}
                for n, info in presence.items() if info["t"] >= cutoff
            ]}})
        if "PleaseSendPeers" in m:
            out.append({"Peers": {"peers": list(peers)[:20]}})
        if "PleaseAlwaysReturnThisMessage" in m:
            pat = m["PleaseAlwaysReturnThisMessage"]
            their_token = pat.get("cookie") if isinstance(pat, dict) else None

    if out:
        if their_token is not None:
            out.append({"AlwaysReturned": {"cookie": their_token}})
        # Include our token in every reply so an unverified peer can echo it back
        # and receive full responses starting from the very next exchange.
        out.append({"PleaseAlwaysReturnThisMessage": {"cookie": token_for(src)}})

        if not is_verified:
            payload = json.dumps(out).encode()
            if len(payload) > 2.5 * req_bytes:
                # Minimum: our token (lets them bootstrap verification) + at most 1 peer.
                # AlwaysReturned is dropped — we are not proving ourselves, just giving
                # them what they need to prove themselves next time.
                trimmed = [{"PleaseAlwaysReturnThisMessage": {"cookie": token_for(src)}}]
                for item in out:
                    if "Peers" in item:
                        trimmed.append({"Peers": {"peers": item["Peers"]["peers"][:1]}})
                        break
                out = trimmed
        sock.sendto(json.dumps(out).encode(), addr)
